src/coreclr/vm/callstubgenerator.cpp

```diff
@@ -2451,8 +2453,176 @@ void CallStubGenerator::TerminateCurrentRoutineIfNotOfNewType(RoutineType type,
+            case CorInfoCallConvExtension::FastcallMemberFunction:
+                unmanagedThisCallConv = true;
+                break;
+            default:
+                break;
+        }
+    }
+
+#if defined(TARGET_WINDOWS)
+    // On these platforms, when making a ThisCall, or other call using a C++ MemberFunction calling convention,
+    // the "this" pointer is passed in the first argument slot.
+    bool rewriteReturnTypeToForceRetBuf = false;
+    if (unmanagedThisCallConv)
+    {
+        rewriteMetaSigFromExplicitThisToHasThis = true;
+        // Also, any struct type other than a few special cases is returned via return buffer for unmanaged calls
+        CorElementType retType = sig.GetReturnType();
+        sig.Reset();
+
+        if (retType == ELEMENT_TYPE_VALUETYPE)
+        {
+            TypeHandle thRetType = sig.GetRetTypeHandleThrowing();
+            MethodTable* pMTRetType = thRetType.AsMethodTable();
+
+            if (pMTRetType->GetInternalCorElementType() == ELEMENT_TYPE_VALUETYPE && !isNativePrimitiveStructType(pMTRetType))
+            {
+                rewriteReturnTypeToForceRetBuf = true;
+            }
+        }
+    }
+#endif // defined(TARGET_WINDOWS)
+
+    // Rewrite ExplicitThis to HasThis. This allows us to use ArgIterator which is unaware of ExplicitThis
+    // in the places where it is needed such as computation of return buffers.
+    if (sig.GetCallingConventionInfo() & IMAGE_CEE_CS_CALLCONV_EXPLICITTHIS)
+    {
+        printf("Managed ExplicitThis to HasThis conversion needed\n");
+        rewriteMetaSigFromExplicitThisToHasThis = true;
+    }
+
+    SigBuilder sigBuilder;
+    if (rewriteMetaSigFromExplicitThisToHasThis)
+    {
+        printf("Rewriting ExplicitThis to implicit this\n");
+        sigBuilder.AppendByte(IMAGE_CEE_CS_CALLCONV_DEFAULT_HASTHIS);
+        if ((sig.NumFixedArgs() == 0) || (sig.HasThis() && !sig.HasExplicitThis()))
+        {
+            ThrowHR(COR_E_BADIMAGEFORMAT);
+        }
+        sigBuilder.AppendData(sig.NumFixedArgs() - 1);
+        TypeHandle thRetType = sig.GetRetTypeHandleThrowing();
+#if defined(TARGET_WINDOWS)
+        if (rewriteReturnTypeToForceRetBuf)
+        {
+            // Change the return type to type large enough it will always need to be returned via return buffer
+            thRetType = CoreLibBinder::GetClass(CLASS__STACKFRAMEITERATOR);
+            _ASSERTE(thRetType.IsValueType());
+            _ASSERTE(thRetType.GetSize() > 64);
+            sigBuilder.AppendElementType(ELEMENT_TYPE_INTERNAL);
+            sigBuilder.AppendPointer(thRetType.AsPtr());
+        }
+        else
+#endif
+        {
+            SigPointer pReturn = sig.GetReturnProps();
+            pReturn.ConvertToInternalExactlyOne(sig.GetModule(), sig.GetSigTypeContext(), &sigBuilder);
+        }
+
+        // Skip the explicit this argument
+        sig.NextArg();
+
+    // Copy rest of the arguments
+        sig.NextArg();
```

Comment:
I think that second sig.NextArg() causes to drop the first real argument after this.

Reply:
I don't believe so. It makes the GetArgProps ready to pick up the arg, although I agree the api is inscrutable. I'll add a comment.